src/generic_actor.cr

```diff
@@ -1,28 +1,72 @@
 require "log"
+require "mutex"
 
 module GenericActor
-  VERSION = "0.1.0"
+  VERSION = "0.1.1"
 
-  @message_loop_started = Atomic::Flag.new
+  private enum ActorState
+    PENDING
+    PROCESSING
+    STOPPED
+  end
+
+  class AbuseStoppedActorException < Exception
+  end
+
+  @message_loop_state = ActorState::PENDING
+  @message_loop_state_mutex = Mutex.new
   @message_queue = Channel(Message).new(100)
+  @was_stop_action_performed = Atomic::Flag.new
 
-  private def check_message_loop
-    return unless @message_loop_started.test_and_set
-    spawn { actor_loop }
+  def stop_actor
+    @message_loop_state_mutex.synchronize do
+      perform_stop_action if @message_loop_state == ActorState::PENDING
+      @message_loop_state = ActorState::STOPPED
+      @message_queue.close
+    end
+  end
+
+  protected def process_stop
+    # You can override process_stop to close resources you want
   end
-  
+
+  private def perform_stop_action
+    return unless @was_stop_action_performed.test_and_set
+    begin
+      process_stop
+    rescue e
+      Log.error {"Cought exception from process_stop: #{e}"}
+    end
+  end
+
   private abstract struct Message
   end
 
+  private def check_message_loop
+    @message_loop_state_mutex.synchronize do
+      case @message_loop_state
+      when ActorState::PROCESSING
+        return
+      when ActorState::STOPPED
+        raise AbuseStoppedActorException.new
+      else
+        spawn { actor_loop }
+        @message_loop_state = ActorState::PROCESSING
+      end
+    end
```

Comment:
```suggestion
    @message_loop_state_mutex.synchronize do
      case @message_loop_state.get
      when  .processing?
        return
      when .stopped?
        raise AbuseStoppedActorException.new
      else
         @message_loop_state.set(:processing) # but better idea compare_and_set to control flow
         spawn { actor_loop }
      end
    end
```

Reply:
I feared a race condition between @message_loop_state.get and @message_loop_state.set(:processing). compare_and_set will help to avoid double spawn. Thank you